sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java

```diff
@@ -144,6 +171,8 @@ public static void main(Function<String, String> environmentVarGetter) throws Ex
     }
 
     System.out.format("Pipeline options %s%n", pipelineOptionsJson);
+    pipelineOptionsJson = removeNestedKey(pipelineOptionsJson, "impersonateServiceAccount");
```

Comment:
OK I now realize why this one is doing a weird thing to the string of the JSON, which is not a good pattern.

You _could_ have just set the pipeline option to `null` after parsing it. But of course then you would have to take a dependency on GCP pipeline options when the SDK harness should not have a dependency on GCP.

So it is a hack. What should really happen is the Dataflow service strips this option before sending the options to the worker. Maybe what you actually want is for the Dataflow pipeline to have two sets of options in the job submission: (1) the options used to submit the job (2) the options used to run the job.

Anyhow, please keep a bug open for this and make an extensive comment here. Because even though you didn't add a dependency on the GCP library, you did actually add a dependency on a GCP concept, plus a hack that someone who comes later will wonder "what is going on here?"

Reply:
My comment sounds too negative - I mean to say that I am OK with this but we need to track the follow-up, make a clear comment, and handle it in Dataflow better.

Reply:
agree with you. I just opened #30301 for some future improvements.